sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -0,0 +1,236 @@
+    val minCmp = ordering.compare(value, minInToType)
+    val maxCmp = ordering.compare(value, maxInToType)
+
+    if (maxCmp > 0) {
+      exp match {
+        case EqualTo(_, _) | GreaterThan(_, _) | GreaterThanOrEqual(_, _) =>
+          falseIfNotNull(fromExp)
+        case LessThan(_, _) | LessThanOrEqual(_, _) =>
+          trueIfNotNull(fromExp)
+        // make sure the expression is evaluated if it is non-deterministic
+        case EqualNullSafe(_, _) if exp.deterministic =>
+          FalseLiteral
+        case _ => exp
+      }
+    } else if (maxCmp == 0) {
+      exp match {
+        case GreaterThan(_, _) =>
+          falseIfNotNull(fromExp)
+        case LessThanOrEqual(_, _) =>
+          trueIfNotNull(fromExp)
+        case LessThan(_, _) =>
+          Not(EqualTo(fromExp, Literal(max, fromType)))
+        case GreaterThanOrEqual(_, _) | EqualTo(_, _) =>
+          EqualTo(fromExp, Literal(max, fromType))
+        case EqualNullSafe(_, _) =>
+          EqualNullSafe(fromExp, Literal(max, fromType))
+        case _ => exp
+      }
+    } else if (minCmp < 0) {
+      exp match {
+        case GreaterThan(_, _) | GreaterThanOrEqual(_, _) =>
+          trueIfNotNull(fromExp)
+        case LessThan(_, _) | LessThanOrEqual(_, _) | EqualTo(_, _) =>
+          falseIfNotNull(fromExp)
+        // make sure the expression is evaluated if it is non-deterministic
+        case EqualNullSafe(_, _) if exp.deterministic =>
+          FalseLiteral
+        case _ => exp
+      }
+    } else if (minCmp == 0) {
+      exp match {
+        case LessThan(_, _) =>
+          falseIfNotNull(fromExp)
+        case GreaterThanOrEqual(_, _) =>
+          trueIfNotNull(fromExp)
+        case GreaterThan(_, _) =>
+          Not(EqualTo(fromExp, Literal(min, fromType)))
+        case LessThanOrEqual(_, _) | EqualTo(_, _) =>
+          EqualTo(fromExp, Literal(min, fromType))
+        case EqualNullSafe(_, _) =>
+          EqualNullSafe(fromExp, Literal(min, fromType))
+        case _ => exp
+      }
+    } else {
+      // This means `value` is within range `(min, max)`. Optimize this by moving the cast to the
+      // literal side.
+      val lit = Cast(Literal(value), fromType)
+      exp match {
+        case GreaterThan(_, _) => GreaterThan(fromExp, lit)
+        case GreaterThanOrEqual(_, _) => GreaterThanOrEqual(fromExp, lit)
+        case EqualTo(_, _) => EqualTo(fromExp, lit)
+        case EqualNullSafe(_, _) => EqualNullSafe(fromExp, lit)
+        case LessThan(_, _) => LessThan(fromExp, lit)
+        case LessThanOrEqual(_, _) => LessThanOrEqual(fromExp, lit)
+        case _ => exp
+      }
+    }
+  }
+
+  /**
+   * Check if the input `fromExp` can be safely cast to `toType` without any loss of precision,
+   * i.e., the conversion is injective. Note this only handles the case when both sides are of
+   * integral type.
+   */
+  private def canImplicitlyCast(fromExp: Expression, toType: DataType,
```

Comment:
indentation nit:
```
def func(
    para1: T,
    para2: T,
    para3: T): R = ...
```

Reply:
BTW I'll also check `!from.foldable`, otherwise it's simpler to not run this rule.

Reply:
Ya. +1 for `foldable` checking.

Reply:
Thanks. I'll check `foldable` in the follow-up PR to handle more types. And also fix the indentation.